pkg/analyzer/nodes/node.go

```diff
@@ -0,0 +1,211 @@
+				"relative_path": relPath,
+			}).Error("extract file")
+			return err
+		}
+		if f != nil {
+			n.files[path] = f
+		}
+		return nil
+	})
+}
+
+// filesIsNotAccessible returns true if the error means the file cannot be read.
+func filesIsNotAccessible(err error) bool {
+	return errors.Is(err, fs.ErrNotExist) || errors.Is(err, fs.ErrPermission)
+}
+
+// extractFile extracts data from the given directory entry, if granted by the
+// path matcher, and if it passes safety checks. Otherwise, returns nil.
+func extractFile(path string, entry fs.DirEntry, pathMatcher matcher.Matcher, m *metrics.FileExtractionMetrics) (*fileMetadata, error) {
+	// Ignore all directories.
+	if entry.IsDir() {
+		return nil, nil
+	}
+	fileInfo, err := entry.Info()
+	if err != nil {
+		if filesIsNotAccessible(err) {
+			m.InaccessibleCount()
+			return nil, nil
+		}
+		// We assume that other errors reflect underlying problems with the filesystem
+		// worth failing the whole extraction.
+		return nil, err
+	}
+	m.FileCount()
+	// FIXME No matcher actually uses the file contents, except the matcher for
+	//       language components, which is not currently used by directory scans, hence we
+	//       offer nil.  This will break if a language component matcher (analyzer) is used.
+	matches, isExtractable := pathMatcher.Match(path, fileInfo, nil)
+	if !matches {
+		return nil, nil
+	}
+	m.MatchCount()
+	// File size limit check.
+	if isExtractable && fileInfo.Size() > analyzer.GetMaxExtractableFileSize() {
+		logrus.Errorf("skipping file %q (%d bytes): size is greater than maxExtractableFileSize of %d MiB",
+			path, fileInfo.Size(), analyzer.GetMaxExtractableFileSize()/1024/1024)
+		return nil, nil
+	}
+	return &fileMetadata{
+		isExecutable:  matcher.IsFileExecutable(fileInfo),
+		isExtractable: isExtractable,
+	}, nil
+}
+
+// Get implements analyzer.Files
+func (n *filesMap) Get(path string) (analyzer.FileData, bool) {
+	// When a previous read error has happened, we act as if all the files map is
+	// empty. Analyzer are expected to handle it gracefully.
+	if f, ok := n.files[path]; n.readError == nil && ok {
+		fileData := analyzer.FileData{Executable: f.isExecutable}
+		if !f.isExtractable {
+			return fileData, true
+		}
+		fileData.Contents, n.readError = os.ReadFile(filepath.Join(n.root, filepath.FromSlash(path)))
+		if n.readError == nil {
+			return fileData, true
+		}
+	}
+	return analyzer.FileData{}, false
+}
+
+// GetFilesPrefix implements analyzer.Files
+func (n *filesMap) GetFilesPrefix(prefix string) (filesMap map[string]analyzer.FileData) {
+	filesMap = make(map[string]analyzer.FileData)
+	for name := range n.files {
+		if strings.HasPrefix(name, prefix) && name != prefix {
+			data, exists := n.Get(name)
+			if exists {
+				filesMap[name] = data
+			}
+		}
+	}
+	return
+}
+
+// readErr returns the first error encountered when reading file content, and
+// resets the error, enabling re-scans using the same map.
+func (n *filesMap) readErr() error {
```

Comment:
who calls this?

Reply:
Just after `clair.DetectFromFiles()` (now `clair.DetectComponents()`), to verify if there was any error when reading files during detection.